src/lib/SlushIO.ml

```diff
@@ -0,0 +1,666 @@
+
+    | `A [`String "GoalProj"; tm1] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      ret @@ GoalProj(tm1)
+
+    | `A [`String "Coe"; tm1; tm2; tm3; tm4] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      ret @@ Coe(tm1, tm2, tm3, tm4)
+
+    | `A [`String "HCom"; tm1; tm2; tm3; tm4; tm5] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ HCom(tm1, tm2, tm3, tm4,tm5)
+
+    | `A [`String "Com"; tm1; tm2; tm3; tm4; tm5] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ Com(tm1, tm2, tm3, tm4,tm5)
+
+    | `A [`String "SubIn"; tm1] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      ret @@ SubIn(tm1)
+
+    | `A [`String "SubOut"; tm1] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      ret @@ SubOut(tm1)
+
+    | `String "Dim0" -> ret @@ Dim0
+    | `String "Dim1" -> ret @@ Dim1
+
+    | `A [`String "Cof"; c] ->
+      cof_of_json c >>= fun c ->
+      ret @@ Cof(c)
+
+    | `A [`String "ForallCof"; tm] ->
+      tm_of_json tm >>= fun tm ->
+      ret @@ ForallCof(tm)
+
+    | `A [`String "CofSplit"; tm] -> raise Todo
+
+    | `String "Prf" -> ret @@ Prf
+
+    | `A [`String "ElIn"; tm] ->
+      tm_of_json tm >>= fun tm ->
+      ret @@ ElIn(tm)
+
+    | `A [`String "ElOut"; tm] ->
+      tm_of_json tm >>= fun tm ->
+      ret @@ ElOut(tm)
+
+    | `A [`String "Box"; tm1; tm2; tm3; tm4; tm5] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ Box(tm1, tm2, tm3, tm4,tm5)
+
+    | `A [`String "Cap"; tm1; tm2; tm3; tm4; tm5] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ Cap(tm1, tm2, tm3, tm4,tm5)
+
+    | `A [`String "VIn"; tm1; tm2; tm3; tm4] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      ret @@ VIn(tm1, tm2, tm3, tm4)
+
+    | `A [`String "VProj"; tm1; tm2; tm3; tm4; tm5] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      tm_of_json tm2 >>= fun tm2 ->
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ VProj(tm1, tm2, tm3, tm4, tm5)
+
+    | `A [`String "CodeExt"; i; tm1; glo; tm2] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      global_of_json glo >>= fun glo ->
+      tm_of_json tm2 >>= fun tm2 ->
+      ret @@ CodeExt(int_of_json i, tm1, glo, tm2) (* todo *)
```

Comment:
Just a small stylistic comment, in ML languages we think of the argument to such a constructor as a tuple --- so we prefer to write `CodeExt (...)` with a space.

Reply:
do you have feelings about single argument constructors? e.g. `ElIn (tm)` (uniform but redundant) vs `ElIn(tm)` (unlikely) or `ElIn tm` (concise but not uniform)?

Reply:
Since most languages don’t have a type of unary tuples , so I think we are safe with the latter haha

Reply:
got it. i shall endeavor to be more jon-friendly in this respect!